### PR TITLE
Avoid enqueuing the same cluster twice

### PR DESCRIPTION
**Pull request by alvaroaleman**
`handler.EnqueueRequestForOwner` enqueues the owning object with the
triggering objects namespace, which doesn't work if the owning object
is non-namespaced.

We worked around this by removing the namespace from clusters in the
`Reconcile`. The problem is that we still have two possible queue keys
for the cluster, allowing two workers to concurrently work on the same
cluster, causing a lot of errors.

This PR changes the whole logic to not use the ownerreference but the
namespace of the objects to find the corresponding cluster.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

/cherrypick release/v2.10
/assign @mrIncompetent

### CODE REVIEW

File: api/pkg/controller/cluster/cluster_controller.go [hunk 1]
```diff
@@ -144,7 +144,20 @@ func Add(
 		return err
 	}
 
-	ownerHandler := &handler.EnqueueRequestForOwner{IsController: true, OwnerType: &kubermaticv1.Cluster{}}
+	dynamicClient := mgr.GetClient()
+	enqueueClusterForNamespacedObject := &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+		clusterList := &kubermaticv1.ClusterList{}
+		if err := dynamicClient.List(context.Background(), &ctrlruntimeclient.ListOptions{}, clusterList); err != nil {
+			// TODO: Is there a better way to handle errors that occur here?
+			glog.Errorf("failed to list Clusters: %v", err)
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/cluster_controller.go]: Typically `utilruntime.HandleErr`.
It executes 2 thinks typically:
- Log the error using glog
- Increase our global error metric

File: api/pkg/controller/cluster/cluster_controller.go [hunk 2]
```diff
@@ -144,7 +144,20 @@ func Add(
 		return err
 	}
 
-	ownerHandler := &handler.EnqueueRequestForOwner{IsController: true, OwnerType: &kubermaticv1.Cluster{}}
+	dynamicClient := mgr.GetClient()
+	enqueueClusterForNamespacedObject := &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/cluster_controller.go]: Can we extract the handler for reuse?

**Comment by alvaroaleman** [comment]
> Can we extract the handler for reuse?

So much code saved :sparkles:

File: api/pkg/controller/util/util.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// EnqueueClusterForNamespacedObject enqueues the cluster that owns a namespaced object, if any
+// It is used by various controllers to react to changes in the resources in the cluster namespace
+func EnqueueClusterForNamespacedObject(client ctrlruntimeclient.Client) *handler.EnqueueRequestsFromMapFunc {
+	return &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+		clusterList := &kubermaticv1.ClusterList{}
+		if err := client.List(context.Background(), &ctrlruntimeclient.ListOptions{}, clusterList); err != nil {
+			utilruntime.HandleError(fmt.Errorf("failed to list Clusters: %v", err))
```
Comment (mrIncompetent) [inline api/pkg/controller/util/util.go:22]: return here

**Comment by alvaroaleman** [comment]
@mrIncompetent Updated, PTAL

**Pull request merged**